plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftParquetPageSource.java

```diff
@@ -0,0 +1,133 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.redshift;
+
+import com.google.common.io.Closer;
+import io.trino.parquet.ParquetCorruptionException;
+import io.trino.parquet.reader.ParquetReader;
+import io.trino.spi.Page;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorPageSource;
+import io.trino.spi.metrics.Metrics;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.OptionalLong;
+
+import static io.trino.plugin.base.util.Closables.closeAllSuppress;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_PARQUET_BAD_DATA;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_PARQUET_CURSOR_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftParquetPageSource
+        implements ConnectorPageSource
+{
+    private final ParquetReader parquetReader;
+    private final Closer closer;
+    private boolean closed;
+    private long completedPositions;
+
+    public RedshiftParquetPageSource(ParquetReader parquetReader, Closeable closeable)
+    {
+        this.parquetReader = requireNonNull(parquetReader, "parquetReader is null");
+        closer = Closer.create();
+        closer.register(requireNonNull(closeable, "closeable is null"));
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return parquetReader.getDataSource().getReadBytes();
+    }
+
+    @Override
+    public OptionalLong getCompletedPositions()
+    {
+        return OptionalLong.of(completedPositions);
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return parquetReader.getDataSource().getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page page;
+        try {
+            page = parquetReader.nextPage();
+        }
+        catch (IOException | RuntimeException e) {
+            closeAllSuppress(e, this);
+            throw handleException(e);
+        }
+
+        if (closed || page == null) {
+            close();
+            return null;
+        }
+
+        completedPositions += page.getPositionCount();
+        return page;
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        return parquetReader.getMemoryContext().getBytes();
+    }
+
+    @Override
+    public void close()
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+
+        try {
+            parquetReader.close();
+            closer.close();
```

Comment:
Its problematic to delete the file here for FTE mode.
In FTE the scheduler is allowed to retry tasks or even schedule another run of the same task on a different node if the current node stopped responding to coordinator but is still active.
Instead of deleting here, all the unloaded files should be deleted in `ConnectorSplitSource#close`

Reply:
Good to know. However, I see that RedshiftUnloadSplitSource#close is called before RedshiftPageSourceProvider#createPageSource locally hence not making suggested changes. Please advise if I misunderstood anything.

Reply:
You can implement `io.trino.plugin.jdbc.JdbcQueryEventListener#cleanupQuery` to run `io.trino.filesystem.TrinoFileSystem#deleteDirectory` on `unloadLocation + "/" + queryId`

Reply:
Nice, Implementing JdbcQueryEventListener  works as expected.